kge/util/distribution-shift/custom_distribution.py

```diff
@@ -0,0 +1,88 @@
+import torch
+import random
+import os
+import distribution_shift_utils as dsutils
+
+
+class CustomDistribution:
+    def __init__(self, checkpoint_path, noise_magnitude):
+        _, self.dataset = dsutils.load_model_and_dataset(checkpoint_path)
+        self.noise_magnitude = noise_magnitude
+
+    def calculate_target_distribution(self, triples):
+        """Calculate a slightly shifted target distribution by adding random noise."""
+
+        relation_triples = dsutils.group_triples(triples, group_type="relation")
+        original_distribution = {relation: 1.0 for relation, _ in relation_triples.items()}
+        
```

Comment:
Target Distribution for Testing:
The target distribution you used to test might be too far from the training distribution. After you reminded me that sampling a sufficient amount of triples isn’t always feasible, I realized we should avoid creating a test set with a large distribution shift. Instead, we can evaluate whether a very small distribution shift affects performance or increases variance first.
A more convenient way to define the target distribution would be as follows:

    Use the original i.i.d. distribution and add a small random noise to the vectors (e.g., randomly increase or decrease values by a small amount).
    This approach will also scale better for large distribution vectors (e.g., with lengths of 30k+), where manual input isn’t practical.